test2/faces40/javaPage/src/main/java/com/sun/faces/test/faces40/javapages/Hello.java

```diff
@@ -0,0 +1,62 @@
+package com.sun.faces.test.faces40.javapages;
+
+import static jakarta.faces.application.StateManager.IS_BUILDING_INITIAL_STATE;
+
+import java.io.IOException;
+import java.util.List;
+
+import jakarta.enterprise.context.ApplicationScoped;
+import jakarta.faces.annotation.FacesViewId;
+import jakarta.faces.component.UIComponent;
+import jakarta.faces.component.UIOutput;
+import jakarta.faces.component.html.HtmlBody;
+import jakarta.faces.component.html.HtmlCommandButton;
+import jakarta.faces.component.html.HtmlForm;
+import jakarta.faces.component.html.HtmlOutputText;
+import jakarta.faces.context.FacesContext;
+import jakarta.faces.view.facelets.Facelet;
+
+@FacesViewId("/hello.xhtml")
+@ApplicationScoped
+public class Hello extends Facelet {
+
+    @Override
+    public void apply(FacesContext facesContext, UIComponent root) throws IOException {
+        if (!facesContext.getAttributes().containsKey(IS_BUILDING_INITIAL_STATE)) {
```

Comment:
Shouldn't this be `ComponentHandler.isNew(root)`?

Reply:
Nope, that one only checks if the component has a parent. This here has to do with the fact the Facelets compiler actually builds the view twice, and then the second time it deletes and adds components.  The difference between the first and second time is the `IS_BUILDING_INITIAL_STATE` constant being set.

At some point we may need to look at this Facelets internal weirdness, but that's far beyond the scope of this issue.

Reply:
Ok. Agreed